### ISSUE TITLE
Feature Request: Reduce the Redundancy of Masterwork Hashes

### ISSUE THREAD

**Issue by Xalavar** [Bungie-net/api#772]
Whenever I look up a weapon's stats and I want to check the masterworks it can have, it shows all of the hashes for possible masterwork tiers (which is either 1 through 5 or 1 through 10) for each possible stat it can improve. Due to this, some weapons can populate up to 100 different hashes that can make it incredibly tedious to sift through.

It would be much more efficient if all the different hashes that represent each of these tiers were condensed into one hash for each stat, and these new hashes would include the minimum and maximum possible tiers for each stat that can be masterworked.

**Comment by vthornheart-bng** [comment]
Oh, are you talking on the definition level?  Or is it returning you all of them in the live data?

**Comment by LordNazo** [comment]
Yeah, I'm referring to the definition. (Sorry, I keep forgetting to mention that.)

**Comment by vthornheart-bng** [comment]
It's okay, no worries!

In this situation, the problem is that because it's a definition it actually can have all of those: and because they're represented as Sockets, what you're seeing is the actual way that those sockets are represented in the data (and thus the way that we need to refer to them - for instance, we have to have the exact index into that array when we attempt to insert the next level plug).

We could build additional helper data specifically for masterworks that would give info specifically for it, but we have a couple of problems (likely the same ones you're running into):

1) The relationship between tiers of masterworks plugs are loosely coupled, such that any code we wrote to break them into tiers would be very brittle and susceptible to breaking as content changes.

2) If the underlying concept of masterworks ever changes again (as it has a few times in its iterations already - 3 times now?), we run the risk of not only breaking the process that builds the data but also of the data structure itself no longer being sufficient, and thus requiring us to make a breaking/non-backward-compatible change to the data.

Because of the history of us being burned by both of these, I tend to lean pretty heavily toward not building artificial constructs for concepts that are purely driven by content.  If masterworks tiers were actually supported by code rather than content, I'd feel a bit safer... but as it is, they exist in a real danger zone for us (the danger zone of someone deciding to change how they work entirely without having to actually pay an engineering cost to do so).

However, I totally recognize that this sucks for you guys as an experience.  I'm going to mark this as an enhancement and give it a think: perhaps there's a more generic concept I could expose that would let me "group" nodes by whatever coupling I find between them, without creating the definitions such that it can only work with a specific concept like masterwork plugs.  I think that might actually work, but I need to think through the implications of such a choice.  It'd mean still taking the cost of (1), but if we can avoid (2) I'd take it.
